### PR TITLE
Update Qnn EP example code to reflect the QNN cache improvement in Ort 

### PR DESCRIPTION
**Pull request by HectorSVC**
Update Qnn EP example code to reflect the QNN cache improvement in Ort https://github.com/microsoft/onnxruntime/pull/17757

### CODE REVIEW

File: c_cxx/QNN_EP/mobilenetv2_classification/README.md [hunk 1]
```diff
@@ -1,17 +1,28 @@
 ## About
-- Builds the sample compiled against the ONNX Runtime built with support for Qualcomm AI Engine Direct SDK (Qualcomm Neural Network (QNN) SDK)
-- The sample uses the QNN EP, run with Qnn CPU banckend and HTP backend
+- Builds the sample compiled against the ONNX Runtime built with support for Qualcomm AI Engine Direct SDK (Qualcomm Neural Network (QNN) SDK).
+- The sample uses the QNN EP, run the float32 model on Qnn CPU banckend.
```
Comment (ivberg) [inline c_cxx/QNN_EP/mobilenetv2_classification/README.md:3]: Maybe this should be a list? "This sample uses the QNN EP to:"
1. foo
2. bar

File: c_cxx/QNN_EP/mobilenetv2_classification/README.md [hunk 2]
```diff
@@ -1,17 +1,28 @@
...
+- The sample uses the QNN EP, run the QDQ model on HTP backend with qnn_context_cache_enable=1, and generates the Onnx model which has QNN context binary embeded.
+- The sample uses the QNN EP, run the QNN context bianry model generated from ONNX Runtime (previous step) on HTP backend, to improve the model initialization time and reduce memory overhead.
```
Comment (ivberg) [inline c_cxx/QNN_EP/mobilenetv2_classification/README.md:5]: bianry spelled wrong here and next bullet

File: c_cxx/QNN_EP/mobilenetv2_classification/README.md [hunk 3]
```diff
@@ -1,17 +1,28 @@
...
 - The sample downloads the mobilenetv2 model from Onnx model zoo, and use mobilenetv2_helper.py to quantize the float32 model to QDQ model which is required for HTP backend
 - The sample is targeted to run on QC ARM64 device.
 - There are 2 ways to improve the session creation time by using of QNN context binary:
-  - Option 1: Use contexty binary generated by OnnxRuntime QNN EP. OnnxRuntime QNN EP use QNN API to generate the QNN context binary, and also dumps some metadata (model name, version, graph meta id, etc) to identify the model. You can just simply set qnn_context_cache_enable to 1 and run with QDQ model. The first run will generate the context binary (Default file name is model_file_name.onnx.bin if qnn_context_cache_path is not set.). The model run afterwards will load from the generated context binary file.
+  - Option 1: Use contexty binary generated by OnnxRuntime QNN EP. OnnxRuntime QNN EP use QNN API to generate the QNN context binary, and also dumps some metadata (model name, version, graph meta id, etc) to identify the model.
```
Comment (ivberg) [inline c_cxx/QNN_EP/mobilenetv2_classification/README.md:10]: spelling - contexty

File: c_cxx/QNN_EP/mobilenetv2_classification/README.md [hunk 4]
```diff
@@ -1,17 +1,28 @@
...
-    - The sample provides mobilenetv2-12_net.json and context binary file libmobilenetv2-12.serialized.bin as exmple (generated by QNN version 2.10). Please follow QNN document to gnereated QNN model_net.json and the context binary file.
+    - The sample also demonstrates the feature to create an Onnx model file from QNN generated context binary file libmobilenetv2-12.serialized.bin to better support customer application migration from native QNN to OnnxRuntime QNN EP. A script [gen_qnn_ctx_onnx_model.py](https://github.com/microsoft/onnxruntime/tree/main/onnxruntime/python/tools/qnn/gen_qnn_ctx_onnx_model.py) is provided to gnereate an Onnx model from QNN generated context binary file. It requires the QNN generated context binary file libmobilenetv2-12.serialized.bin and pre-converted QNN mobilenetv2-12_net.json.
```
Comment (ivberg) [inline c_cxx/QNN_EP/mobilenetv2_classification/README.md:16]: Spelling - gnereate

File: c_cxx/QNN_EP/mobilenetv2_classification/README.md [hunk 5]
```diff
@@ -1,17 +1,28 @@
...
+    - a. Convert model to QNN model and gnereate the QNN context binary file. The sample provides mobilenetv2-12_net.json and context binary file libmobilenetv2-12.serialized.bin as exmple (generated by QNN version 2.10). Please follow QNN document to gnereated QNN model_net.json and the context binary file.
```
Comment (ivberg) [inline c_cxx/QNN_EP/mobilenetv2_classification/README.md:17]: Spelling - gnereate

File: c_cxx/QNN_EP/mobilenetv2_classification/run_qnn_ep_sample.bat [hunk 6]
```diff
@@ -89,22 +87,24 @@ copy /y %ORT_BIN%\QnnHtpV68Stub.dll .
 copy /y %ORT_BIN%\QnnSystem.dll .
 copy /y %ORT_BIN%\libQnnHtpV68Skel.so .
```
Comment (ivberg) [inline c_cxx/QNN_EP/mobilenetv2_classification/run_qnn_ep_sample.bat:88]: Per my [PR](https://github.com/microsoft/onnxruntime-inference-examples/pull/307) do we also want/need v73 from QNN SDK now? I am not sure myself but assume it's support for a newer chipset

copy /y %ORT_BIN%\QnnHtpV73Stub.dll
copy /y %ORT_BIN%\libQnnHtpV73Skel.so

File: c_cxx/QNN_EP/mobilenetv2_classification/main.cpp [hunk 7]
```diff
@@ -252,39 +233,53 @@ void run_ort_qnn_ep(std::string backend, std::string input_path, std::string qnn
...
 void PrintHelp() {
   std::cout << "To run the sample, use the following command:" << std::endl;
-  std::cout << "Example: ./qnn_ep_sample --cpu <path_to_raw_input>" << std::endl;
-  std::cout << "To Run with QNN CPU backend. Example: ./qnn_ep_sample --cpu kitten_input.raw" << std::endl;
-  std::cout << "To Run with QNN HTP backend. Example: ./qnn_ep_sample --htp kitten_input.raw" << std::endl;
-  std::cout << "To Run with QNN native context binary on QNN HTP backend . Example: ./qnn_ep_sample --qnn kitten_input_nhwc.raw" << std::endl;
+  std::cout << "Example: ./qnn_ep_sample --cpu <model_path> <path_to_raw_input>" << std::endl;
+  std::cout << "To Run with QNN CPU backend. Example: ./qnn_ep_sample --cpu mobilenetv2-12_shape.onnx kitten_input.raw" << std::endl;
+  std::cout << "To Run with QNN HTP backend. Example: ./qnn_ep_sample --htp mobilenetv2-12_quant_shape.onnx kitten_input.raw" << std::endl;
+  std::cout << "To Run with QNN HTP backend and generate Qnn context binary model. Example: ./qnn_ep_sample --htp mobilenetv2-12_quant_shape.onnx kitten_input.raw --gen_ctx" << std::endl;
```
Comment (aungthetnaing) [inline c_cxx/QNN_EP/mobilenetv2_classification/main.cpp:239]: any error check possible if onnx file created with --gen-ctx runs with CPU EP?

File: c_cxx/QNN_EP/mobilenetv2_classification/README.md [hunk 8]
```diff
@@ -1,17 +1,28 @@
...
       - Example command used:
         - qnn-onnx-converter --input_list ./input.txt --input_network ./mobilenetv2-12.onnx --output_path ./mobilenetv2-12.cpp -b 1 --bias_bw 32
         - qnn-model-lib-generator -c ./mobilenetv2-12.cpp -b ./mobilenetv2-12.bin -o ./mobilenetv2_classification/qnn_lib
         - qnn-context-binary-generator --backend ${QNN_SDK_ROOT}/target/x86_64-linux-clang/lib/libQnnHtp.so --model ./mobilenetv2_classification/qnn_lib/x86_64-linux-clang/libmobilenetv2-12.so --binary_file libmobilenetv2-12.serialized
+    - b. Create an Onnx model file from QNN generated context binary file libmobilenetv2-12.serialized.bin. More details refer to run_qnn_ep_sample.bat & [gen_qnn_ctx_onnx_model.py](https://github.com/microsoft/onnxruntime/tree/main/onnxruntime/python/tools/qnn/gen_qnn_ctx_onnx_model.py)
```
[the remark this answers is not in the document]
Comment (HectorSVC) [reply c_cxx/QNN_EP/mobilenetv2_classification/README.md:22]: Good point. I need to investigate this. Need to try in Ort QNN EP first. We can get this PR merged first. I'll create a separate task to track this.

---
In reply to: [1353427093](https://github.com/microsoft/onnxruntime-inference-examples/pull/309#discussion_r1353427093) [](ancestors = 1353427093)

**Review by aungthetnaing — APPROVED** [review]

**Review by ivberg — APPROVED** [review]

**Review by adrianlizarraga — APPROVED** [review]

**Pull request merged**